// ngx-cleave-directive/bundles/ngx-cleave-directive.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('tslib'), require('@angular/core'), require('@angular/forms'), require('cleave.js')) :
    typeof define === 'function' && define.amd ? define('ngx-cleave-directive', ['exports', 'tslib', '@angular/core', '@angular/forms', 'cleave.js'], factory) :
    (factory((global['ngx-cleave-directive'] = {}),global.tslib,global.ng.core,global.ng.forms,null));
}(this, (function (exports,tslib_1,core,forms,Cleave) { 'use strict';

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */
    var NgxCleaveDirective = (function () {
        function NgxCleaveDirective(elementRef, valueAccessors) {
            this.elementRef = elementRef;
            this.valueAccessors = valueAccessors;
        }
        Object.defineProperty(NgxCleaveDirective.prototype, "cleave", {
            set: /**
             * @param {?} cleave
             * @return {?}
             */ function (cleave) {
                this._cleave = cleave;
                this.setCleave();
            },
            enumerable: true,
            configurable: true
        });
        /**
         * @return {?}
         */
        NgxCleaveDirective.prototype.ngOnInit = /**
         * @return {?}
         */
            function () {
                var _this = this;
                if (!this.valueAccessors) {
                    console.warn('Note: The cleave directive should be used with the ngModel, formControl or formControlName directives.');
                    return;
                }
                if (this.valueAccessors.length) {
                    this._valueAccessor = this.valueAccessors[0];
                    this._writeValue = this._valueAccessor.writeValue;
                    this._valueAccessor.writeValue = function (value) {
                        if (_this._writeValue) {
                            _this._writeValue.call(_this._valueAccessor, value);
                        }
                        _this.setCleave();
                    };
                }
            };
        /**
         * @return {?}
         */
        NgxCleaveDirective.prototype.ngOnDestroy = /**
         * @return {?}
         */
            function () {
                if (this._valueAccessor && this._writeValue) {
                    this._valueAccessor.writeValue = this._writeValue;
                }
                if (this._cleaveInstance) {
                    this._cleaveInstance.destroy();
                }
            };
        /**
         * @param {?} value
         * @return {?}
         */
        NgxCleaveDirective.prototype.onInput = /**
         * @param {?} value
         * @return {?}
         */
            function (value) {
                this._value = value;
            };
        /**
         * @return {?}
         */
        NgxCleaveDirective.prototype.setCleave = /**
         * @return {?}
         */
            function () {
                var _this = this;
                if (this._cleaveInstance) {
                    this._cleaveInstance.destroy();
                }
                var /** @type {?} */ el = this.elementRef.nativeElement;
                this._cleaveInstance = new Cleave(el, tslib_1.__assign({}, this._cleave, { onValueChanged: function (_a) {
                        var target = _a.target;
                        if (target.value !== _this._value) {
                            _this.dispatchEvent(el, 'input');
                        }
                    } }));
                // hack for model -> view cleave
                setTimeout(function () { return _this.dispatchEvent(el, 'input'); }, 0);
            };
        /**
         * @param {?} el
         * @param {?} eventType
         * @return {?}
         */
        NgxCleaveDirective.prototype.dispatchEvent = /**
         * @param {?} el
         * @param {?} eventType
         * @return {?}
         */
            function (el, eventType) {
                var /** @type {?} */ event = document.createEvent('Event');
                event.initEvent(eventType, false, false);
                el.dispatchEvent(event);
            };
        NgxCleaveDirective.decorators = [
            { type: core.Directive, args: [{
                        selector: 'input[cleave],textarea[cleave]',
                    },] },
        ];
        /** @nocollapse */
        NgxCleaveDirective.ctorParameters = function () {
            return [
                { type: core.ElementRef, },
                { type: Array, decorators: [{ type: core.Inject, args: [forms.NG_VALUE_ACCESSOR,] }, { type: core.Optional },] },
            ];
        };
        NgxCleaveDirective.propDecorators = {
            "cleave": [{ type: core.Input },],
            "onInput": [{ type: core.HostListener, args: ['input', ['$event.target.value'],] },],
        };
        return NgxCleaveDirective;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */
    var NgxCleaveDirectiveModule = (function () {
        function NgxCleaveDirectiveModule() {
        }
        NgxCleaveDirectiveModule.decorators = [
            { type: core.NgModule, args: [{
                        imports: [],
                        declarations: [NgxCleaveDirective],
                        exports: [NgxCleaveDirective],
                    },] },
        ];
        return NgxCleaveDirectiveModule;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */

    exports.NgxCleaveDirective = NgxCleaveDirective;
    exports.NgxCleaveDirectiveModule = NgxCleaveDirectiveModule;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
