// ngx-cleave-directive/esm5/ngx-cleave.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Directive, ElementRef, HostListener, Inject, Input, Optional, } from '@angular/core';
import { NG_VALUE_ACCESSOR, } from '@angular/forms';
import * as Cleave from 'cleave.js';
var NgxCleaveDirective = /** @class */ (function () {
    function NgxCleaveDirective(elementRef, valueAccessors) {
        this.elementRef = elementRef;
        this.valueAccessors = valueAccessors;
    }
    Object.defineProperty(NgxCleaveDirective.prototype, "cleave", {
        set: /**
         * @param {?} cleave
         * @return {?}
         */
        function (cleave) {
            this._cleave = cleave;
            this.setCleave();
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    NgxCleaveDirective.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        if (!this.valueAccessors) {
            console.warn('Note: The cleave directive should be used with the ngModel, formControl or formControlName directives.');
            return;
        }
        if (this.valueAccessors.length) {
            this._valueAccessor = this.valueAccessors[0];
            this._writeValue = this._valueAccessor.writeValue;
            this._valueAccessor.writeValue = function (value) {
                if (_this._writeValue) {
                    _this._writeValue.call(_this._valueAccessor, value);
                }
                _this.setCleave();
            };
        }
    };
    /**
     * @return {?}
     */
    NgxCleaveDirective.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        if (this._valueAccessor && this._writeValue) {
            this._valueAccessor.writeValue = this._writeValue;
        }
        if (this._cleaveInstance) {
            this._cleaveInstance.destroy();
        }
    };
    /**
     * @param {?} value
     * @return {?}
     */
    NgxCleaveDirective.prototype.onInput = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        this._value = value;
    };
    /**
     * @return {?}
     */
    NgxCleaveDirective.prototype.setCleave = /**
     * @return {?}
     */
    function () {
        var _this = this;
        if (this._cleaveInstance) {
            this._cleaveInstance.destroy();
        }
        var /** @type {?} */ el = this.elementRef.nativeElement;
        this._cleaveInstance = new Cleave(el, tslib_1.__assign({}, this._cleave, { onValueChanged: function (_a) {
                var target = _a.target;
                if (target.value !== _this._value) {
                    _this.dispatchEvent(el, 'input');
                }
            } }));
        // hack for model -> view cleave
        setTimeout(function () { return _this.dispatchEvent(el, 'input'); }, 0);
    };
    /**
     * @param {?} el
     * @param {?} eventType
     * @return {?}
     */
    NgxCleaveDirective.prototype.dispatchEvent = /**
     * @param {?} el
     * @param {?} eventType
     * @return {?}
     */
    function (el, eventType) {
        var /** @type {?} */ event = document.createEvent('Event');
        event.initEvent(eventType, false, false);
        el.dispatchEvent(event);
    };
    NgxCleaveDirective.decorators = [
        { type: Directive, args: [{
                    selector: 'input[cleave],textarea[cleave]',
                },] },
    ];
    /** @nocollapse */
    NgxCleaveDirective.ctorParameters = function () { return [
        { type: ElementRef, },
        { type: Array, decorators: [{ type: Inject, args: [NG_VALUE_ACCESSOR,] }, { type: Optional },] },
    ]; };
    NgxCleaveDirective.propDecorators = {
        "cleave": [{ type: Input },],
        "onInput": [{ type: HostListener, args: ['input', ['$event.target.value'],] },],
    };
    return NgxCleaveDirective;
}());
export { NgxCleaveDirective };
function NgxCleaveDirective_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    NgxCleaveDirective.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    NgxCleaveDirective.ctorParameters;
    /** @type {!Object<string,!Array<{type: !Function, args: (undefined|!Array<?>)}>>} */
    NgxCleaveDirective.propDecorators;
    /** @type {?} */
    NgxCleaveDirective.prototype._cleave;
    /** @type {?} */
    NgxCleaveDirective.prototype._cleaveInstance;
    /** @type {?} */
    NgxCleaveDirective.prototype._value;
    /** @type {?} */
    NgxCleaveDirective.prototype._valueAccessor;
    /** @type {?} */
    NgxCleaveDirective.prototype._writeValue;
    /** @type {?} */
    NgxCleaveDirective.prototype.elementRef;
    /** @type {?} */
    NgxCleaveDirective.prototype.valueAccessors;
}
