// ngx-cleave-directive/fesm2015/ngx-cleave-directive.js
import { Directive, ElementRef, HostListener, Inject, Input, Optional, NgModule } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import * as Cleave from 'cleave.js';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
class NgxCleaveDirective {
    /**
     * @param {?} elementRef
     * @param {?} valueAccessors
     */
    constructor(elementRef, valueAccessors) {
        this.elementRef = elementRef;
        this.valueAccessors = valueAccessors;
    }
    /**
     * @param {?} cleave
     * @return {?}
     */
    set cleave(cleave) {
        this._cleave = cleave;
        this.setCleave();
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        if (!this.valueAccessors) {
            console.warn('Note: The cleave directive should be used with the ngModel, formControl or formControlName directives.');
            return;
        }
        if (this.valueAccessors.length) {
            this._valueAccessor = this.valueAccessors[0];
            this._writeValue = this._valueAccessor.writeValue;
            this._valueAccessor.writeValue = (value) => {
                if (this._writeValue) {
                    this._writeValue.call(this._valueAccessor, value);
                }
                this.setCleave();
            };
        }
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        if (this._valueAccessor && this._writeValue) {
            this._valueAccessor.writeValue = this._writeValue;
        }
        if (this._cleaveInstance) {
            this._cleaveInstance.destroy();
        }
    }
    /**
     * @param {?} value
     * @return {?}
     */
    onInput(value) {
        this._value = value;
    }
    /**
     * @return {?}
     */
    setCleave() {
        if (this._cleaveInstance) {
            this._cleaveInstance.destroy();
        }
        const /** @type {?} */ el = this.elementRef.nativeElement;
        this._cleaveInstance = new Cleave(el, Object.assign({}, this._cleave, { onValueChanged: ({ target }) => {
                if (target.value !== this._value) {
                    this.dispatchEvent(el, 'input');
                }
            } }));
        // hack for model -> view cleave
        setTimeout(() => this.dispatchEvent(el, 'input'), 0);
    }
    /**
     * @param {?} el
     * @param {?} eventType
     * @return {?}
     */
    dispatchEvent(el, eventType) {
        const /** @type {?} */ event = document.createEvent('Event');
        event.initEvent(eventType, false, false);
        el.dispatchEvent(event);
    }
}
NgxCleaveDirective.decorators = [
    { type: Directive, args: [{
                selector: 'input[cleave],textarea[cleave]',
            },] },
];
/** @nocollapse */
NgxCleaveDirective.ctorParameters = () => [
    { type: ElementRef, },
    { type: Array, decorators: [{ type: Inject, args: [NG_VALUE_ACCESSOR,] }, { type: Optional },] },
];
NgxCleaveDirective.propDecorators = {
    "cleave": [{ type: Input },],
    "onInput": [{ type: HostListener, args: ['input', ['$event.target.value'],] },],
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
class NgxCleaveDirectiveModule {
}
NgxCleaveDirectiveModule.decorators = [
    { type: NgModule, args: [{
                imports: [],
                declarations: [NgxCleaveDirective],
                exports: [NgxCleaveDirective],
            },] },
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */

export { NgxCleaveDirective, NgxCleaveDirectiveModule };
