// ngx-cleave-directive/fesm5/ngx-cleave-directive.js
import { __assign } from 'tslib';
import { Directive, ElementRef, HostListener, Inject, Input, Optional, NgModule } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import * as Cleave from 'cleave.js';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var NgxCleaveDirective = /** @class */ (function () {
    function NgxCleaveDirective(elementRef, valueAccessors) {
        this.elementRef = elementRef;
        this.valueAccessors = valueAccessors;
    }
    Object.defineProperty(NgxCleaveDirective.prototype, "cleave", {
        set: /**
         * @param {?} cleave
         * @return {?}
         */
        function (cleave) {
            this._cleave = cleave;
            this.setCleave();
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    NgxCleaveDirective.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        if (!this.valueAccessors) {
            console.warn('Note: The cleave directive should be used with the ngModel, formControl or formControlName directives.');
            return;
        }
        if (this.valueAccessors.length) {
            this._valueAccessor = this.valueAccessors[0];
            this._writeValue = this._valueAccessor.writeValue;
            this._valueAccessor.writeValue = function (value) {
                if (_this._writeValue) {
                    _this._writeValue.call(_this._valueAccessor, value);
                }
                _this.setCleave();
            };
        }
    };
    /**
     * @return {?}
     */
    NgxCleaveDirective.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        if (this._valueAccessor && this._writeValue) {
            this._valueAccessor.writeValue = this._writeValue;
        }
        if (this._cleaveInstance) {
            this._cleaveInstance.destroy();
        }
    };
    /**
     * @param {?} value
     * @return {?}
     */
    NgxCleaveDirective.prototype.onInput = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        this._value = value;
    };
    /**
     * @return {?}
     */
    NgxCleaveDirective.prototype.setCleave = /**
     * @return {?}
     */
    function () {
        var _this = this;
        if (this._cleaveInstance) {
            this._cleaveInstance.destroy();
        }
        var /** @type {?} */ el = this.elementRef.nativeElement;
        this._cleaveInstance = new Cleave(el, __assign({}, this._cleave, { onValueChanged: function (_a) {
                var target = _a.target;
                if (target.value !== _this._value) {
                    _this.dispatchEvent(el, 'input');
                }
            } }));
        // hack for model -> view cleave
        setTimeout(function () { return _this.dispatchEvent(el, 'input'); }, 0);
    };
    /**
     * @param {?} el
     * @param {?} eventType
     * @return {?}
     */
    NgxCleaveDirective.prototype.dispatchEvent = /**
     * @param {?} el
     * @param {?} eventType
     * @return {?}
     */
    function (el, eventType) {
        var /** @type {?} */ event = document.createEvent('Event');
        event.initEvent(eventType, false, false);
        el.dispatchEvent(event);
    };
    NgxCleaveDirective.decorators = [
        { type: Directive, args: [{
                    selector: 'input[cleave],textarea[cleave]',
                },] },
    ];
    /** @nocollapse */
    NgxCleaveDirective.ctorParameters = function () { return [
        { type: ElementRef, },
        { type: Array, decorators: [{ type: Inject, args: [NG_VALUE_ACCESSOR,] }, { type: Optional },] },
    ]; };
    NgxCleaveDirective.propDecorators = {
        "cleave": [{ type: Input },],
        "onInput": [{ type: HostListener, args: ['input', ['$event.target.value'],] },],
    };
    return NgxCleaveDirective;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var NgxCleaveDirectiveModule = /** @class */ (function () {
    function NgxCleaveDirectiveModule() {
    }
    NgxCleaveDirectiveModule.decorators = [
        { type: NgModule, args: [{
                    imports: [],
                    declarations: [NgxCleaveDirective],
                    exports: [NgxCleaveDirective],
                },] },
    ];
    return NgxCleaveDirectiveModule;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */

export { NgxCleaveDirective, NgxCleaveDirectiveModule };
